src/primary_names.lua

```diff
@@ -0,0 +1,301 @@
+local utils = require("utils")
+local primaryNames = {}
+
+local PRIMARY_NAME_COST = 100000000 -- 100 IO
+
+---@alias WalletAddress string
+---@alias ArNSName string
+
+---@class PrimaryNames
+---@field owners table<WalletAddress, PrimaryName> - map indexed by owner address containing the primary name and all metadata, used for reverse lookups
+---@field names table<ArNSName, WalletAddress> - map indexed by primary name containing the owner address, used for reverse lookups
+---@field requests table<WalletAddress, PrimaryNameRequest> - map indexed by owner address containing the request, used for pruning expired requests
+
+PrimaryNames = PrimaryNames or {
+	requests = {},
+	names = {},
+	owners = {},
+}
+
+---@class PrimaryName
+---@field name ArNSName
+---@field baseName ArNSName
+---@field startTimestamp number
+
+---@class PrimaryNameWithOwner
+---@field name ArNSName
+---@field baseName ArNSName
+---@field owner WalletAddress
+---@field startTimestamp number
+
+---@class PrimaryNameRequest
+---@field name ArNSName -- the name being requested
+---@field baseName ArNSName -- the base name, identified when creating the name request
+---@field initiator WalletAddress -- the process id that made the request
+---@field startTimestamp number -- the timestamp of the request
+---@field endTimestamp number -- the timestamp of the request expiration
+
+--- Creates a transient request for a primary name. This is done by a user and must be approved by the name owner of the base name.
+--- @param name string -- the name being requested, this could be an undername provided by the ant
+--- @param initiator string -- the address that is creating the primary name request, e.g. the ANT process id
+--- @param timestamp number -- the timestamp of the request
+--- @return PrimaryNameRequest|PrimaryNameWithOwner primaryNameRequestOrWithOwner - the request created, or the primary name with owner data if the request is approved
+function primaryNames.createPrimaryNameRequest(name, initiator, timestamp)
+	local baseName = name:match("[^_]+$") or name
+
+	--- existing request for primary name from wallet?
+	local existingRequest = primaryNames.getPrimaryNameRequest(name)
+	assert(not existingRequest, "Primary name request for '" .. name .. "' already exists") -- TODO: should we error here or just let them create a new request and pay the fee again?
+
+	--- check the primary name is not already owned
+	local primaryNameOwner = primaryNames.getAddressForPrimaryName(name)
+	assert(not primaryNameOwner, "Primary name is already owned")
+
+	local record = arns.getRecord(baseName)
+	assert(record, "ArNS record '" .. baseName .. "' does not exist")
+
+	--- TODO: replace with funding plan
+	assert(balances.walletHasSufficientBalance(initiator, PRIMARY_NAME_COST), "Insufficient balance")
+
+	--- transfer the primary name cost from the initiator to the treasury
+	balances.transfer(ao.id, initiator, PRIMARY_NAME_COST)
+
+	local request = {
+		name = name,
```

Comment:
The base name is always derivable from the name, so I'd lobby that we excise the baseName field from here if possible in order to save on process memory and trade off for the compute to get the base name any time it's needed.

Reply:
Done.